### PR TITLE
Get test_grammar passing

### PR DESCRIPTION
**Pull request by slide**
Most fixes were merged in from ironpython2:

https://github.com/IronLanguages/ironpython2/pull/297
https://github.com/IronLanguages/ironpython2/pull/296
https://github.com/IronLanguages/ironpython2/pull/294

### CODE REVIEW

File: Src/IronPython/Compiler/Parser.cs [hunk 1]
```diff
@@ -1167,6 +1172,9 @@ public class Parser : IDisposable { // TODO: remove IDisposable
 
             ret.Body = body;
             ret.HeaderIndex = rEnd;
+            if (annotation != null) {
```
Comment (slozier) [inline Src/IronPython/Compiler/Parser.cs]: Null check needed?

File: Src/IronPython/Compiler/Ast/FunctionDefinition.cs [hunk 2]
```diff
@@ -155,6 +155,26 @@ public FunctionDefinition(string name, Parameter[] parameters, Statement body, S
             internal set { _decorators = value; }
         }
 
+        public Expression ReturnAnnotation {
+            get; internal set;
+        }
+
+        public IDictionary<string, Expression> Annotations {
+            get {
+                var annotations = new Dictionary<string, Expression>();
```
Comment (slozier) [inline Src/IronPython/Compiler/Ast/FunctionDefinition.cs]: Creates a dict's on every call?

File: Src/IronPython/Compiler/Ast/FunctionDefinition.cs [hunk 3]
```diff
@@ -492,17 +557,24 @@ class FunctionDefinitionInstruction : Instruction {
...
 
                 _context = context;
                 _defaultCount = defaultCount;
                 _def = definition;
                 _name = name;
+                _annotationCount = annotationCount;
             }
 
             public override int Run(InterpretedFrame frame) {
+                PythonDictionary annotations = new PythonDictionary();
```
Comment (slozier) [inline Src/IronPython/Compiler/Ast/FunctionDefinition.cs]: Create a dict on every invokation?

File: Src/IronPython/Compiler/Ast/Parameter.cs [hunk 4]
```diff
@@ -64,10 +63,7 @@ public Parameter(string name)
             get { return _name; }
         }
 
-        public Expression Annotation {
-            get { return _annotation; }
-            set { _annotation = value; }
-        }
+        public Expression Annotation { get; set; } = null;
```
Comment (slozier) [inline Src/IronPython/Compiler/Ast/Parameter.cs]: `null` assignment is redundant.

File: Src/IronPython/Runtime/PythonFunction.cs [hunk 5]
```diff
@@ -106,14 +108,15 @@ public PythonFunction(CodeContext context, FunctionCode code, PythonDictionary g
             _compat = CalculatedCachedCompat();
         }
 
-        internal PythonFunction(CodeContext/*!*/ context, FunctionCode funcInfo, object modName, object[] defaults, MutableTuple closure) {
+        internal PythonFunction(CodeContext/*!*/ context, FunctionCode funcInfo, object modName, object[] defaults, MutableTuple closure, PythonDictionary annotations) {
```
Comment (slozier) [inline Src/IronPython/Runtime/PythonFunction.cs]: nit: Maybe put `annotations` before `closure`?

**Comment by slozier** [comment]
Can you also include the generator script for `TokenKind.Generated`?
Since this covers most of IronLanguages/ironpython2#294 I would also include the test.

File: Src/IronPython/Runtime/PythonFunction.cs (hunk 5 above)
Comment (slide) [reply Src/IronPython/Runtime/PythonFunction.cs]: Sure, I can do that. It's internal, so it shouldn't affect anything outside.

**Comment by slide** [comment]
Yes, I'll add that as well.

**Pull request merged**